File: src/lib/ir.ts
export type Dialect = 'Postgres' | 'MySql' | 'Mssql' | 'Oracle';

export interface ConnectionDef {
	id: string;
	name: string;
	dialect: Dialect;
	host: string;
	port: number;
	database: string;
	username: string;
}

export interface ColumnInfo {
	name: string;
	data_type: string;
	nullable: boolean;
	is_indexed: boolean;
}

export interface IndexInfo {
	name: string;
	columns: string[];
	is_unique: boolean;
	is_primary: boolean;
}

export interface TableInfo {
	schema: string;
	name: string;
	columns: ColumnInfo[];
	indexes: IndexInfo[];
}

export interface Schema {
	tables: TableInfo[];
}

export interface ConnectionWithSchema {
	def: ConnectionDef;
	schema: Schema | null;
}

export const DIALECTS: { value: Dialect; label: string; defaultPort: number }[] = [
	{ value: 'Postgres', label: 'PostgreSQL', defaultPort: 5432 },
	{ value: 'MySql', label: 'MySQL', defaultPort: 3306 },
	{ value: 'Mssql', label: 'SQL Server', defaultPort: 1433 },
	{ value: 'Oracle', label: 'Oracle', defaultPort: 1521 }
];

export function qualifiedName(table: TableInfo): string {
	return table.schema ? `${table.schema}.${table.name}` : table.name;
}

export const CURRENT_SCHEMA_VERSION = 2;

export type FilterOp =
	| 'Eq'
	| 'Ne'
	| 'Gt'
	| 'Gte'
	| 'Lt'
	| 'Lte'
	| 'Like'
	| 'NotLike'
	| 'Ilike'
	| 'In'
	| 'NotIn'
	| 'Between'
	| 'IsNull'
	| 'IsNotNull'
	| 'IsEmpty'
	| 'IsNotEmpty';

export type GroupConnector = 'And' | 'Or';

export type LiteralKind = 'Text' | 'Int' | 'Float' | 'Bool' | 'Date' | 'DateTime';

export interface TableRef {
	schema: string;
	name: string;
	alias: string;
}

export interface ColumnSel {
	table_alias: string;
	column: string;
}

export interface JoinSpec {
	left_alias: string;
	left_column: string;
	right_alias: string;
	right_column: string;
}

export interface FilterLiteral {
	kind: LiteralKind;
	text: string;
}

export type FilterValue =
	| { Single: FilterLiteral }
	| { List: FilterLiteral[] }
	| { Pair: [FilterLiteral, FilterLiteral] };

export interface FilterSpec {
	table_alias: string;
	column: string;
	op: FilterOp;
	value: FilterValue | null;
}

export interface FilterGroup {
	connector: GroupConnector;
	children: FilterNode[];
}

export type FilterNode =
	| { Leaf: FilterSpec }
	| { Group: FilterGroup };

export interface QuerySpec {
	tables: TableRef[];
	columns: ColumnSel[];
	joins: JoinSpec[];
	filters: FilterGroup;
	limit: number;
	schema_version: number;
}

export interface QueryResult {
	columns: string[];
	rows: JsonValue[][];
	row_count: number;
	truncated: boolean;
	warnings: string[];
}

export type JsonValue = string | number | boolean | null | { [key: string]: JsonValue } | JsonValue[];

export type ValueKind = 'None' | 'Single' | 'List' | 'Pair';

export function valueKindForOp(op: FilterOp): ValueKind {
	switch (op) {
		case 'IsNull':
		case 'IsNotNull':
		case 'IsEmpty':
		case 'IsNotEmpty':
			return 'None';
		case 'In':
		case 'NotIn':
			return 'List';
		case 'Between':
			return 'Pair';
		default:
			return 'Single';
	}
}

export function needsValue(op: FilterOp): boolean {
	return valueKindForOp(op) !== 'None';
}

export const FILTER_OPS: { value: FilterOp; label: string }[] = [
	{ value: 'Eq', label: '=' },
	{ value: 'Ne', label: '≠' },
	{ value: 'Gt', label: '>' },
	{ value: 'Gte', label: '≥' },
	{ value: 'Lt', label: '<' },
	{ value: 'Lte', label: '≤' },
	{ value: 'Like', label: 'like' },
	{ value: 'NotLike', label: 'not like' },
	{ value: 'Ilike', label: 'ilike' },
	{ value: 'In', label: 'in' },
	{ value: 'NotIn', label: 'not in' },
	{ value: 'Between', label: 'between' },
	{ value: 'IsNull', label: 'is null' },
	{ value: 'IsNotNull', label: 'is not null' },
	{ value: 'IsEmpty', label: 'is empty' },
	{ value: 'IsNotEmpty', label: 'is not empty' }
];

export type ColumnCategory = 'Text' | 'Numeric' | 'Bool' | 'Date' | 'DateTime' | 'Other';

export function classifyColumn(dataType: string): ColumnCategory {
	const dt = dataType.toLowerCase();
	if (dt === 'bool' || dt === 'boolean' || dt === 'bit') return 'Bool';
	if (dt === 'date') return 'Date';
	if (dt.startsWith('timestamp') || dt.startsWith('datetime') || dt === 'datetime2' || dt === 'smalldatetime')
		return 'DateTime';
	if (
		[
			'int', 'integer', 'smallint', 'bigint', 'mediumint', 'tinyint',
			'serial', 'bigserial', 'decimal', 'numeric', 'real', 'double',
			'float', 'float4', 'float8', 'money', 'smallmoney', 'double precision'
		].includes(dt) ||
		dt.startsWith('decimal') ||
		dt.startsWith('numeric')
	)
		return 'Numeric';
	if (
		[
			'text', 'varchar', 'char', 'character', 'character varying', 'string',
			'tinytext', 'mediumtext', 'longtext', 'nvarchar', 'nchar',
			'varchar2', 'nvarchar2', 'clob', 'nclob', 'xml'
		].includes(dt) ||
		dt.startsWith('varchar') ||
		dt.startsWith('char') ||
		dt.startsWith('nchar') ||
		dt.startsWith('nvarchar')
	)
		return 'Text';
	return 'Other';
}

export function literalKindForColumn(dataType: string): LiteralKind {
	switch (classifyColumn(dataType)) {
		case 'Numeric':
			return 'Int';
		case 'Bool':
			return 'Bool';
		case 'Date':
			return 'Date';
		case 'DateTime':
			return 'DateTime';
		default:
			return 'Text';
	}
}

export function opsForColumn(dataType: string): FilterOp[] {
	switch (classifyColumn(dataType)) {
		case 'Text':
			return ['Eq', 'Ne', 'Like', 'NotLike', 'Ilike', 'In', 'NotIn', 'IsNull', 'IsNotNull', 'IsEmpty', 'IsNotEmpty'];
		case 'Numeric':
			return ['Eq', 'Ne', 'Gt', 'Gte', 'Lt', 'Lte', 'In', 'NotIn', 'Between', 'IsNull', 'IsNotNull'];
		case 'Bool':
			return ['Eq', 'Ne', 'IsNull', 'IsNotNull'];
		case 'Date':
		case 'DateTime':
			return ['Eq', 'Ne', 'Gt', 'Gte', 'Lt', 'Lte', 'Between', 'IsNull', 'IsNotNull'];
		default:
			return ['Eq', 'Ne', 'IsNull', 'IsNotNull'];
	}
}

export function opLabel(op: FilterOp): string {
	return FILTER_OPS.find((f) => f.value === op)?.label ?? op;
}

export function defaultFilterGroup(): FilterGroup {
	return { connector: 'And', children: [] };
}

export function makeLiteral(dataType: string, text: string): FilterLiteral {
	return { kind: literalKindForColumn(dataType), text };
}

export function makeFilter(
	tableAlias: string,
	column: string,
	dataType: string,
	op: FilterOp,
	text: string
): FilterSpec {
	const kind = literalKindForColumn(dataType);
	const vk = valueKindForOp(op);
	let value: FilterValue | null = null;
	if (vk === 'Single') {
		value = { Single: { kind, text } };
	} else if (vk === 'List') {
		value = { List: text ? [{ kind, text }] : [] };
	} else if (vk === 'Pair') {
		value = { Pair: [{ kind, text: '' }, { kind, text: '' }] };
	}
	return { table_alias: tableAlias, column, op, value };
}

export const MAX_LIMIT = 1000;
export const DEFAULT_LIMIT = 100;

export interface SavedQuery {
	id: string;
	name: string;
	connection_id: string;
	spec: QuerySpec;
	created_at: string;
}

export interface HistoryEntry {
	id: string;
	connection_id: string;
	connection_name: string;
	spec: QuerySpec;
	row_count: number;
	warnings: string[];
	error: Option<string>;
	timestamp: string;
}

type Option<T> = T | null;

export interface Settings {
	blocked_schemas: string[];
	explain_cost_threshold: number;
	theme: string;
}

File: src-tauri/tests/ir.rs
use std::collections::BTreeMap;

use safe_db_lib::query::ir::{
    BindValue, CURRENT_SCHEMA_VERSION, ColumnSel, FilterGroup, FilterLiteral, FilterOp,
    GroupConnector, LiteralKind, QuerySpec, TableRef, ValueKind,
};

#[test]
fn filter_op_value_kind_table() {
    assert_eq!(FilterOp::Eq.value_kind(), ValueKind::Single);
    assert_eq!(FilterOp::Ne.value_kind(), ValueKind::Single);
    assert_eq!(FilterOp::Gt.value_kind(), ValueKind::Single);
    assert_eq!(FilterOp::Gte.value_kind(), ValueKind::Single);
    assert_eq!(FilterOp::Lt.value_kind(), ValueKind::Single);
    assert_eq!(FilterOp::Lte.value_kind(), ValueKind::Single);
    assert_eq!(FilterOp::Like.value_kind(), ValueKind::Single);
    assert_eq!(FilterOp::NotLike.value_kind(), ValueKind::Single);
    assert_eq!(FilterOp::Ilike.value_kind(), ValueKind::Single);
    assert_eq!(FilterOp::In.value_kind(), ValueKind::List);
    assert_eq!(FilterOp::NotIn.value_kind(), ValueKind::List);
    assert_eq!(FilterOp::Between.value_kind(), ValueKind::Pair);
    assert_eq!(FilterOp::IsNull.value_kind(), ValueKind::None);
    assert_eq!(FilterOp::IsNotNull.value_kind(), ValueKind::None);
    assert_eq!(FilterOp::IsEmpty.value_kind(), ValueKind::None);
    assert_eq!(FilterOp::IsNotEmpty.value_kind(), ValueKind::None);
}

#[test]
fn needs_value_is_inverse_of_value_kind_none() {
    assert!(!FilterOp::IsNull.needs_value());
    assert!(!FilterOp::IsNotNull.needs_value());
    assert!(!FilterOp::IsEmpty.needs_value());
    assert!(!FilterOp::IsNotEmpty.needs_value());
    assert!(FilterOp::Eq.needs_value());
    assert!(FilterOp::Between.needs_value());
    assert!(FilterOp::In.needs_value());
}

#[test]
fn sql_operator_for_comparison_and_like_ops() {
    assert_eq!(FilterOp::Eq.sql_operator(), Some("="));
    assert_eq!(FilterOp::Ne.sql_operator(), Some("<>"));
    assert_eq!(FilterOp::Gt.sql_operator(), Some(">"));
    assert_eq!(FilterOp::Gte.sql_operator(), Some(">="));
    assert_eq!(FilterOp::Lt.sql_operator(), Some("<"));
    assert_eq!(FilterOp::Lte.sql_operator(), Some("<="));
    assert_eq!(FilterOp::Like.sql_operator(), Some("LIKE"));
    assert_eq!(FilterOp::NotLike.sql_operator(), Some("NOT LIKE"));
    assert_eq!(FilterOp::Ilike.sql_operator(), None);
    assert_eq!(FilterOp::In.sql_operator(), None);
    assert_eq!(FilterOp::Between.sql_operator(), None);
    assert_eq!(FilterOp::IsNull.sql_operator(), None);
}

#[test]
fn bind_value_from_literal_text() {
    let lit = FilterLiteral {
        kind: LiteralKind::Text,
        text: "Alice".to_string(),
    };
    match BindValue::from_literal(&lit).unwrap() {
        BindValue::Text(s) => assert_eq!(s, "Alice"),
        other => panic!("expected Text, got {other:?}"),
    }
}

#[test]
fn bind_value_from_literal_int_ok_and_err() {
    let ok = FilterLiteral {
        kind: LiteralKind::Int,
        text: "42".to_string(),
    };
    match BindValue::from_literal(&ok).unwrap() {
        BindValue::Int(n) => assert_eq!(n, 42),
        other => panic!("expected Int, got {other:?}"),
    }
    let bad = FilterLiteral {
        kind: LiteralKind::Int,
        text: "not a number".to_string(),
    };
    let err = BindValue::from_literal(&bad).unwrap_err();
    assert!(err.contains("not a valid integer"), "got: {err}");
}

#[test]
fn bind_value_from_literal_float_ok_and_err() {
    let ok = FilterLiteral {
        kind: LiteralKind::Float,
        text: "1.5".to_string(),
    };
    match BindValue::from_literal(&ok).unwrap() {
        BindValue::Float(f) => assert!((f - 1.5).abs() < 1e-9),
        other => panic!("expected Float, got {other:?}"),
    }
    let bad = FilterLiteral {
        kind: LiteralKind::Float,
        text: "abc".to_string(),
    };
    let err = BindValue::from_literal(&bad).unwrap_err();
    assert!(err.contains("not a valid number"), "got: {err}");
}

#[test]
fn bind_value_from_literal_decimal_ok_and_err() {
    let ok = FilterLiteral {
        kind: LiteralKind::Decimal,
        text: "12345678901234567890.1234".to_string(),
    };
    match BindValue::from_literal(&ok).unwrap() {
        BindValue::Decimal(n) => assert_eq!(n.to_string(), "12345678901234567890.1234"),
        other => panic!("expected Decimal, got {other:?}"),
    }
    let bad = FilterLiteral {
        kind: LiteralKind::Decimal,
        text: "abc".to_string(),
    };
    let err = BindValue::from_literal(&bad).unwrap_err();
    assert!(err.contains("not a valid decimal"), "got: {err}");
}

#[test]
fn bind_value_from_literal_bool_accepts_canonical_and_aliases() {
    for text in ["true", "TRUE", "True", "1", "yes", "YES"] {
        let lit = FilterLiteral {
            kind: LiteralKind::Bool,
            text: text.to_string(),
        };
        assert!(matches!(
            BindValue::from_literal(&lit).unwrap(),
            BindValue::Bool(true)
        ));
    }
    for text in ["false", "FALSE", "False", "0", "no", "NO", ""] {
        let lit = FilterLiteral {
            kind: LiteralKind::Bool,
            text: text.to_string(),
        };
        assert!(matches!(
            BindValue::from_literal(&lit).unwrap(),
            BindValue::Bool(false)
        ));
    }
    let bad = FilterLiteral {
        kind: LiteralKind::Bool,
        text: "maybe".to_string(),
    };
    let err = BindValue::from_literal(&bad).unwrap_err();
    assert!(err.contains("not a valid boolean"), "got: {err}");
}

#[test]
fn bind_value_from_literal_date_and_datetime_are_typed() {
    let d = FilterLiteral {
        kind: LiteralKind::Date,
        text: "2025-01-02".to_string(),
    };
    match BindValue::from_literal(&d).unwrap() {
        BindValue::Date(date) => assert_eq!(date.to_string(), "2025-01-02"),
        other => panic!("expected Date, got {other:?}"),
    }
    let dt = FilterLiteral {
        kind: LiteralKind::DateTime,
        text: "2025-01-02T03:04:05Z".to_string(),
    };
    match BindValue::from_literal(&dt).unwrap() {
        BindValue::DateTime(datetime) => assert_eq!(datetime.to_string(), "2025-01-02 03:04:05"),
        other => panic!("expected DateTime, got {other:?}"),
    }
}

#[test]
fn filter_group_default_is_empty_and_group_with_id() {
    let g = FilterGroup::default();
    assert!(!g.id.trim().is_empty());
    assert_eq!(g.connector, GroupConnector::And);
    assert!(g.children.is_empty());
}

#[test]
fn query_spec_round_trip_with_connector_overrides() {
    let spec = QuerySpec {
        tables: vec![TableRef {
            schema: "public".into(),
            name: "users".into(),
            alias: "t0".into(),
        }],
        columns: vec![ColumnSel {
            table_alias: "t0".into(),
            column: "id".into(),
        }],
        joins: vec![],
        filters: FilterGroup {
            id: "root".into(),
            connector: GroupConnector::And,
            children: vec![],
        },
        limit: 50,
        schema_version: CURRENT_SCHEMA_VERSION,
        connector_overrides: BTreeMap::from([("leaf-1".to_string(), GroupConnector::Or)]),
    };
    let json = serde_json::to_string(&spec).unwrap();
    let back: QuerySpec = serde_json::from_str(&json).unwrap();
    assert_eq!(back.limit, 50);
    assert_eq!(back.schema_version, CURRENT_SCHEMA_VERSION);
    assert_eq!(
        back.connector_overrides.get("leaf-1"),
        Some(&GroupConnector::Or)
    );
    assert_eq!(back.tables[0].alias, "t0");
}

#[test]
fn query_spec_deserializes_legacy_spec_without_connector_overrides_or_id() {
    let json = r#"{
        "tables": [],
        "columns": [],
        "joins": [],
        "filters": { "connector": "And", "children": [] },
        "limit": 100
    }"#;
    let spec: QuerySpec = serde_json::from_str(json).unwrap();
    assert_eq!(spec.schema_version, CURRENT_SCHEMA_VERSION);
    assert!(spec.connector_overrides.is_empty());
    assert_eq!(spec.filters.id, "");
}
